x/stake/types/validator.go

```diff
@@ -234,145 +299,102 @@ func (v Validator) ABCIValidatorZero() abci.Validator {
 	}
 }
 
-// Status returns the validator's bond status inferred from the pool shares.
-func (v Validator) Status() sdk.BondStatus {
-	return v.PoolShares.Status
-}
-
-// UpdateStatus updates the location of the shares within a validator if it's
-// bond status has changed.
+// UpdateStatus updates the location of the shares within a validator
+// to reflect the new status
 func (v Validator) UpdateStatus(pool Pool, NewStatus sdk.BondStatus) (Validator, Pool) {
-	var tokens int64
 
-	switch v.Status() {
+	switch v.Status {
 	case sdk.Unbonded:
-		if NewStatus == sdk.Unbonded {
+
+		switch NewStatus {
+		case sdk.Unbonded:
 			return v, pool
+		case sdk.Bonded:
+			pool = pool.looseTokensToBonded(v.Tokens)
 		}
-		pool, tokens = pool.removeSharesUnbonded(v.PoolShares.Amount)
-
 	case sdk.Unbonding:
```

Comment:
Do we ever use this status (`sdk.Unbonding`)? I only see it in `validator_test.go` and I'm not clear why we need it in the model; unbonding delegations / redelegations seem to provide the requisite tracking

Also see https://github.com/cosmos/cosmos-sdk/issues/1673

Reply:
Incorrect - we need this status for when an entire validator begins unbonding (with it's delegators still attached)  because it has been revoked - opening an issue to verify that this is happening

Reply:
https://github.com/cosmos/cosmos-sdk/issues/1676